### PR TITLE
[canvaskit] make picture disposal work on older browsers

### PR DESCRIPTION
**Pull request by yjbanov**
Our object cache and resurrection for browsers that don't support `FinalizationRegistry` (which includes current Safari 14) was incompatible with `Picture.dispose`. This fixes it.

### CODE REVIEW

File: lib/web_ui/lib/src/engine/canvaskit/picture.dart [hunk 1]
```diff
@@ -24,14 +30,38 @@ class CkPicture extends ManagedSkiaObject<SkPicture> implements ui.Picture {
...
+  /// 
+  /// This extra flag is necessary on top of [rawSkiaObject] because
+  /// [rawSkiaObject] being null does not indicate permanent deletion. See
+  /// similar flag [SkiaObjectBox.isDeletedPermanently].
+  bool _isDisposed = false;
+
   @override
   void dispose() {
+    if (_isDisposed) {
+      return;
```
Comment (dnfield) [inline lib/web_ui/lib/src/engine/canvaskit/picture.dart:49]: Isn't it a mistake to call dispose more than once? In the framework and on the VM side we generally assert not debug disposed.

File: lib/web_ui/lib/src/engine/canvaskit/picture.dart [hunk 2]
```diff
@@ -24,14 +30,38 @@ class CkPicture extends ManagedSkiaObject<SkPicture> implements ui.Picture {
...
+    }
+    if (Instrumentation.enabled) {
```
Comment (dnfield) [inline lib/web_ui/lib/src/engine/canvaskit/picture.dart:51]: Guard so this doesn't end up in release?

File: lib/web_ui/lib/src/engine/canvaskit/picture.dart [hunk 3]
```diff
@@ -24,14 +30,38 @@ class CkPicture extends ManagedSkiaObject<SkPicture> implements ui.Picture {
...
+  /// This is indended to be used in tests and assertions only.
+  bool get debugIsDisposed => _isDisposed;
+
+  /// This is set to true when [dispose] is called and is never reset back to
+  /// false.
+  /// 
+  /// This extra flag is necessary on top of [rawSkiaObject] because
+  /// [rawSkiaObject] being null does not indicate permanent deletion. See
+  /// similar flag [SkiaObjectBox.isDeletedPermanently].
+  bool _isDisposed = false;
```
Comment (dnfield) [inline lib/web_ui/lib/src/engine/canvaskit/picture.dart:44]: I appreciate the comment, but how would we ever have a null skia object after initializing this?

Comment (yjbanov) [reply lib/web_ui/lib/src/engine/canvaskit/picture.dart:44]: The parent class (`ManagedSkiaObject`) has a `didDelete` method that nulls it out. It is called when an object is evicted from the cache, at which point the Skia object is deleted. That deletion may not be permanent. The object may still be resurrected. OTOH disposal is permanent.

File: lib/web_ui/lib/src/engine/canvaskit/picture.dart (hunk 1 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/canvaskit/picture.dart:49]: Good point. Changed to assert. I didn't see an assert in the AOT version, but I now empirically verified that it does throw an exception.

File: lib/web_ui/lib/src/engine/canvaskit/picture.dart (hunk 2 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/canvaskit/picture.dart:51]: Potentially. Although as of today, the only effect is to eliminate any potentially expensive logic that might be happening in the instrumentation.

**Review by dnfield — APPROVED** [review]

**Pull request merged**